### DISCUSSION TITLE
Personalizzazione delle risposte e proprietà nidificate

### DISCUSSION THREAD

**Discussion by danieleartico** [fattureincloud/api#354] [Q&A] [answered]
Ciao, da una lista di fatture filtrata (`GET /issued_documents?q=...`) ci serve recuperare, per ciascuna fattura:

- Anno (`year`)
- Numero (`number`)
- Data di scadenza (`next_due_date`)
- Indirizzo e-mail del cliente (`entity` → `email`)

Anno, numero e data di scadenza scendono tranquillamente.
L'indirizzo e-mail del cliente, invece, non c'è verso di recuperarlo.

Interrogazioni che abbiamo provato, facendo leva sui parametri `fields` e `fieldset` (`GET /issued_documents/?q=...&fields=...&fieldset=...)`:

1. `fields` e `fieldset` non impostati.
Scendono i campi `basic` della fattura e del cliente. Come da documentazione ([link](https://developers.fattureincloud.it/docs/basics/additional-capabilities/customize-response/#-available-fieldsets)), `entity` → `email` non è presente.
3. `fields` non impostato e `fieldset=detailed`.
Scendono i campi `detailed` della fattura e del cliente ma, a differenza di quanto documentato, `entity` → `email` continua a non essere presente.
4. `fields=year,number,next_due_date,entity` e `fieldset` non impostato.
Scendono esattamente i campi specificati della fattura e tutti quelli `basic` del cliente. Quindi `entity` → `email` non è presente.
5. `fields=year,number,next_due_date,entity` e `fieldset=detailed`.
Come al punto 2. scendono tutti i campi della fattura e del cliente, ma ancora senza `entity` → `email`.

Aggiungo l'ovvio: almeno una parte delle fatture ottenute dall'interrogazione ha sicuramente il cliente con l'e-mail impostata, tantè che interrogando l'endpoint `GET /entities/clients/123?fieldset=detailed` l'indirizzo e-mail viene restituito correttamente.

Ho 2 domande, per favore:

1. È corretto che, richiedendo una lista di fatture con `fieldset=detailed`, le proprietà esposte dei rispettivi clienti (elementi nidificati) non siano tutte quelle annoverate nella documentazione?
2. C'è un modo per specificare una proprietà nidificata in `fields`? Per es. `fields=year,number,next_due_date,entity[email]`, oppure una sintassi equivalente.
Questa naturalmente sarebbe l'opzione ideale perché permetterebbe di alleggerire notevolmente la risposta.

Attualmente, per aggirare il problema otteniamo la lista delle fatture e interroghiamo di nuovo le API per chiedere l'indirizzo e-mail di ciascuno dei clienti, ma concorderete che non sia molto elegante. Per piacere, vi viene in mente una soluzione migliore?

Grazie!

**Comment by emajo** [comment, marked as answer]
Ciao @danieleartico, purtroppo non c'è modo di recuperare i dati del cliente quali email, telefono e fax direttamente dalla chiamata di lista documenti, in quanto quei campi non vengono salvati effettivamente nella fattura, a differenza del resto dei dati dell'entity.
Infatti allo stato attuale passando l'id di un cliente in creazione documento i suoi dati verranno salvati anche all'interno della fattura, in modo che, se i dati del cliente cambiassero in futuro, le fatture emesse in precedenza continuino a presentare i dati del passato.

Questo comportamento però non include i tre campi specificati precedentemente perché effettivamente non fanno parte della risorsa fattura, passando il campo mail in creazione di fattura verrà ignorato per lo stesso motivo, l'unico modo per farla apparire nel pdf è appunto quello di passare l'id del cliente relativo.

Purtroppo sappiamo che non è un comportamento consistente, soprattutto perché come segnalato sulla documentazione il campo email è presente anche se non utilizzabile (questo perché utilizziamo gli stessi modelli per più chiamate).

Grazie,
Emanuele.

Per quanto riguarda il recupero di campi innestati come il campo entity.name non è possibile farlo con il parametro fields che funziona solamente sui capi del primo livello.

**Comment by danieleartico** [comment]
Va bene, ti ringrazio per la risposta velocissima ed esaustiva.
